src/data_structures/matrix.rs

```diff
@@ -0,0 +1,122 @@
+use std::fmt;
+
+/// A two dimensional array
+#[derive(Debug, PartialEq, Eq)]
+pub struct Matrix<T> {
+    cells: Vec<Option<T>>,
+    width: usize,
+    height: usize,
+}
+
+impl<T> Matrix<T>
+where
+    T: Clone,
+{
+    /// Constructs a new `Matrix` with the given width and height.
+    pub fn new(width: usize, height: usize) -> Self {
+        Matrix {
+            cells: vec![None; width * height],
+            width,
+            height,
+        }
+    }
+
+    /// Get the value of the cell at position (x, y). Returns `None` if the position is out of
+    /// bounds or if the cell is empty.
+    pub fn get(&self, x: usize, y: usize) -> Option<&T> {
+        match self.cells.get((y * self.width) + x) {
+            // Out of bounds
+            None => None,
+            // Nothing in the cell
+            Some(None) => None,
+            // Something in the cell
+            Some(value) => value.as_ref(),
+        }
+    }
+
+    /// Sets the cell value at the position (x, y). Panics if the position is out of bounds.
+    pub fn set(&mut self, x: usize, y: usize, value: T) {
+        if y >= self.height || x >= self.width {
+            panic!("Out of bounds");
+        } else {
+            self.cells[(y * self.width) + x] = Some(value);
+        }
+    }
+
+    pub fn width(&self) -> usize {
+        self.width
+    }
+
+    pub fn height(&self) -> usize {
+        self.height
+    }
+}
+
+impl<T> fmt::Display for Matrix<T>
```

Comment:
I get why this is done this way I think but it would still be nice if we could print the grid out in context to the string so our human eyes can check it easier. But that's more then fine.